### PR TITLE
Reorganize code

### PR DESCRIPTION
**Pull request by mathewmoon**

### CODE REVIEW

File: cmd/root.go [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+	"github.com/mathewmoon/tfdoc-html/formatter"
+	"github.com/mathewmoon/tfdoc-html/writers"
+
+	"github.com/spf13/cobra"
+	"github.com/terraform-docs/terraform-docs/format"
+	"github.com/terraform-docs/terraform-docs/print"
+	"github.com/terraform-docs/terraform-docs/terraform"
+)
+
+type Config struct {
```
Comment (abaehre) [inline cmd/root.go]: this should definitely be private.

File: cmd/root.go [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+	file, err := cmd.Flags().GetString("file")
+	if err != nil {
+		exitWithError(err, 1)
+	}
+
+	if len(args) == 0 {
+		exitWithError(errors.New("must provide [PATH] as first argument"), 1)
+	}
+
+	source_path := args[0]
```
Comment (abaehre) [inline cmd/root.go]: nit: you don't need to create a var for this.

File: cmd/root.go [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+
+		if config.s3Uri != "" {
+			_, err := writers.S3Upload(config.s3Uri, doc)
+			if err != nil {
+				exitWithError(err, 2)
+			}
+		}
+
+		if !config.noStdout {
+			fmt.Println(doc)
```
Comment (abaehre) [inline cmd/root.go:127]: normally you use some sort of logger but this probably fine.

File: cmd/root.go [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+		exitWithError(err, 3)
+	}
+}
+
+func exitWithError(msg error, code int) {
+	fmt.Printf("Error: %s ", msg)
+	os.Exit(code)
+}
+
+func init() {
```
Comment (abaehre) [inline cmd/root.go:144]: init's you generally keep at the top. If you "hide" them and people don't find it it is super confusing how these are actually getting setup.

File: formatter/html.go [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+)
+
+/*
+	Converts a string to html
+	Args:
+		`data`: The document body to inject.
+		`css_file`. If `css_file` is anything other than an empty string then we will attempt to read the file
+								 and inject the contents into the style tag
+*/
+func GenerateHtml(data string, css_file string) (string, error) {
```
Comment (abaehre) [inline formatter/html.go]: nit: for same parameter types you can define it just at the end
(data, css_file string)

File: formatter/html.go [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+	extensions := parser.CommonExtensions | parser.AutoHeadingIDs
+	parser := parser.NewWithExtensions(extensions)
+
+	md := []byte(data)
+	body := string(markdown.ToHTML(md, parser, nil))
+
+	var html string
```
Comment (abaehre) [inline formatter/html.go]: nit: generally you define the variable then override it in the if statement to avoid else's.

File: formatter/html.go [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+	} else {
+		buf, err := ioutil.ReadFile(css_file)
+
+		if err != nil {
+			return "", err
+		}
+		style = string(buf)
+
+	}
+	html_template := `
```
Comment (abaehre) [inline formatter/html.go:57]: I feel like this should be a const at the top? but that's a personal preference.

File: writers/s3.go [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+		SharedConfigState: session.SharedConfigEnable,
+	})
+	if err != nil {
+		log.Fatal(err)
+	}
+	return session
+}
+
+/* Parse a fully qualified S3 URI into a Bucket and Key */
+func ParseUri(uri string) (string, string) {
```
Comment (abaehre) [inline writers/s3.go]: should be private.

File: writers/s3.go [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+	if !strings.HasPrefix(uri, "s3://") {
+		fmt.Printf("URI '%s' is not a valid S3 URI", uri)
+		os.Exit(3)
```
Comment (abaehre) [inline writers/s3.go]: generally you want to bubble errors up so the way you handle them is consistent.

File: writers/s3.go [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+	}
+	uri = strings.ReplaceAll(uri, "s3://", "")
+	parts := strings.Split(uri, "/")
+
+	bucket := parts[0]
```
Comment (abaehre) [inline writers/s3.go]: these should have length checks.

File: writers/local.go [hunk 11]
```diff
@@ -0,0 +1,14 @@
+package writers
```
Comment (abaehre) [inline writers/local.go:1]: so generally between packages you want to create a public interface that exposes a function (or two) and then you create structs for each of these that implement that interface (the structs can / should be private). Then based on your config you can create one of those structs, and then put that interface into your config. It'll clean up your logic a bit in root.go as well.

https://gitlab.morningconsult.com/mci/dev-ops/hodor/-/merge_requests/38/diffs

the cache implementation there is a decent example.

File: formatter/html.go [hunk 12]
```diff
@@ -0,0 +1,77 @@
+package formatter
```
Comment (abaehre) [inline formatter/html.go:1]: I noted below about the interface thing. You can do that here if you want to future proof, but if you're only ever going to be generating html then I think this is fine.

**Review by abaehre — CHANGES_REQUESTED** [review]

File: writers/local.go (hunk 11 above)
Comment (mathewmoon) [reply writers/local.go:1]: So do you mean that in the entrypoint you would have something like?:
```
doc := MakeDoc(config)
doc.upload()
doc.print()
doc.markdown
```

File: formatter/html.go (hunk 6 above)
Comment (mathewmoon) [reply formatter/html.go]: Seems inefficient since you write a non empty string to memory twice if css_file != "", but sure.

**Pull request merged**